### ISSUE TITLE
custom model: Didn't find custom op for name 'TFLite_Detection_PostProcess' with version 1

### ISSUE THREAD

**Issue by drakkan** [nnsuite/nnstreamer#916]
I'm trying to use my custom model with nnstreamer, I converted it to tflite format using the instructions here:

https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tensorflowlite.md

when I try to use this model in nnstreamer (I used object_detection example), I get this error:

```
Didn't find custom op for name 'TFLite_Detection_PostProcess' with version 1

Registration failed.

**
ERROR:/build/nnstreamer-gcYSLk/nnstreamer-0.0.3/gst/tensor_filter/tensor_filter_tensorflow_lite.c:101:tflite_open: assertion failed: (retval == 0)
Aborted (core dumped)
```

I think this could be solved updating tensorflow to 1.12 version

the ppa has version 1.9, is nnstreamer tested with 1.12?

**Comment by myungjoo** [comment]
If ```TFLite_Detection_PostProcess``` is not supported in TF 1.9, but only from 1.12, yes, you need to install TF 1.12 in your machine.

We haven't tested TF 1.12 along with nnstreamer.

However, as long as TF keeps backward compatibility (I believe so), you can install TF 1.12 to your machine without rebuilding nnstreamer and run your model.

In the long term (likely around 2019.1Q), we are going to detach neural network framework (caffe, tf, ...) supports from main nnstreamer so that nn-framework support codes can be built independently from nnstreamer.

**Comment by myungjoo** [comment]
Oh.. and one more thing. Even with 1.12, tf-lite may not support that operator.

So, for such models not supporting tf-lite, you may need to wait for tensorflow (not tensorflow-lite) support, which is supposed to be imminent ( @helloahn is working on it )

**Comment by jijoongmoon** [comment]
As far as I know, you may need to implement your own custom layer for the TFLite_Detection_PostProcess. 
When you convert tensorflow model to tensorflow-lite with TOCO, it leaves or ask to implement custom layer to user for the unsupported layer by tflite. So, in this case, TOCO just leave blank layer named "TFLite_DEtection_PostProcess" and you should implement this.

**Comment by helloahn** [comment]
@drakkan 
It's about [`Custom Operator`](https://www.tensorflow.org/lite/custom_operators) of Tensorflow lite.
Usually SSD model calculate the result with their own custom operators.
The 4 outputs has their own roles, like `num_detections`, `detection_classes`, `detection_scores`, `detection_boxes`.
However, to make those outputs, we need to attach custom operator in person and if there is no custom op, it occurs error like your case.

To make it simple, I converted the model without the custom ops like [this page](https://github.com/nnsuite/nnstreamer/wiki/%5BTF-Lite%5D-How-to-Convert-Tensorflow-Model-to-TF-Lite-Model#4-convert-with-toco) for the example.
After that I wrote the code for the post processing(like nms) by myself by using 2 outputs(one for boxes, the other for detected objects).
https://github.com/nnsuite/nnstreamer/blob/e0d610520017207b450fc7a5ceb9ffad7c843a00/nnstreamer_example/example_object_detection/nnstreamer_example_object_detection.cc#L505-L540

Since the examples(`example_filter`,`example_object_detection`) only works with the model which has same in/output tensor format and datatype, your customized model won't work if it has different properties.

If you want to make an example with your model which has different properties rather than our pre-build example, you have to make new one.

**Comment by drakkan** [comment]
Thanks,

I converted the model as you suggested

```
toco --graph_def_file=/tmp/tflite_graph.pb --output_file=/tmp/out1.tflite --input_shapes=1,300,300,3 --input_arrays=Preprocessor/sub --output_arrays='concat,concat_1' --inference_type=FLOAT
```

but now I have this error:

```
tensorflow/contrib/lite/simple_memory_arena.cc:82 it->size != alloc.size (2430000 != 0)
**
ERROR:/build/nnstreamer-DVSBcL/nnstreamer-0.0.3/gst/tensor_filter/tensor_filter_tensorflow_lite.c:101:tflite_open: assertion failed: (retval == 0)
Aborted (core dumped)
```

still something wrong with my model?

**Comment by helloahn** [comment]
@drakkan 
It could be a bug of r1.9 according to the link you leaved.
If you want to run, anyway, could you try [this model?](https://github.com/nnsuite/testcases/tree/master/DeepLearningModels/tensorflow-lite/ssd_mobilenet_v2_coco)
I converted ssd model in person and I executed the example with it.
Furthermore, [This Link](https://github.com/nnsuite/testcases/tree/master/DeepLearningModels/tensorflow/ssd_mobilenet_v2) gives you a pbfile.
You can use the pbfile to tflite file with toco command.

**Comment by drakkan** [comment]
already tested with that models and it works, I'll try to build tensorflow 1.12 to see if this solves the issue with my model (that works fine inside an android app using the android tensorflow demo)

**Comment by helloahn** [comment]
@drakkan If you have no more question, can we close this issue?
